File: desktop/app/utils/fleet_status.py
"""
Canonical fleet effective-status derivation for the desktop (offline mirror of
backend/app/services/fleet_status.py and the normative
shared/fleet_status_reference.py).

ONE core decides "what state is this vehicle in right now" — used by:
  * the DomainStore snapshot build (`compute_fleet_sets`, session adapter)
  * the DomainStore temporal recompute (`compute_fleet_sets_rows`, pure)
  * the Dashboard offline overview (`compute_fleet_counts`)
so the Vehicles page, the Dashboard and the BoundaryClock can never disagree.

Precedence (mutually exclusive):
    SOLD / INACTIVE  (structural)  >  MAINTENANCE  >  RENTED  >  RESERVED  >  AVAILABLE

RENTED is time-derived: a blocking reservation (RESERVED or ACTIVE) whose
window contains ``now`` means the car is physically out — this business has no
separate "pickup" step. RESERVED means a blocking reservation is still
UPCOMING (``now < start``), surfaced so staff see committed demand.

Interval rule: half-open [start, end). A vehicle is occupied for
``start <= now < end``; exactly at ``end`` it is free again. An active
maintenance ticket (status NOT IN COMPLETED/CANCELLED) with no explicit end
occupies its vehicle from start until it is closed (effective_end = +infinity).
"""
from datetime import datetime, timedelta, timezone

from app.utils.datetime_utils import parse_datetime_utc

STRUCTURAL_STATUSES = ("SOLD", "INACTIVE")
TERMINAL_MAINTENANCE_STATUSES = ("COMPLETED", "CANCELLED")
BLOCKING_RESERVATION_STATUSES = ("ACTIVE", "RESERVED")
FAR_FUTURE = datetime(9999, 12, 31, tzinfo=timezone.utc)

EFFECTIVE_AVAILABLE = "AVAILABLE"
EFFECTIVE_RESERVED = "RESERVED"
EFFECTIVE_RENTED = "RENTED"
EFFECTIVE_MAINTENANCE = "MAINTENANCE"


def _get(obj, name, default=None):
    """Read ``name`` from a dict or an ORM object."""
    if isinstance(obj, dict):
        return obj.get(name, default)
    return getattr(obj, name, default)


def _maintenance_end(m):
    end = parse_datetime_utc(
        _get(m, "actual_end_datetime") or _get(m, "expected_end_datetime")
    )
    return end or FAR_FUTURE


# ── pure core (row dicts / ORM objects, explicit `now`) ─────────────────────
def compute_fleet_sets_rows(vehicles, reservations, maintenances, now=None):
    """Return (rented_vids, reserved_vids, maintenance_vids, total_vehicles).

    Works on any iterable of dicts or ORM rows exposing ``id``/``status`` for
    vehicles and ``vehicle_id``/``status``/``start_datetime``/``end_datetime``
    (reservations) resp. ``expected_end_datetime``/``actual_end_datetime``
    (maintenances). This is THE derivation — the session variant below is a
    thin adapter over it.
    """
    now = now or datetime.now(timezone.utc)

    structural = {
        _get(v, "id") for v in vehicles
        if (_get(v, "status") or "").upper() in STRUCTURAL_STATUSES
    }
    total_vehicles = sum(
        1 for v in vehicles
        if (_get(v, "status") or "").upper() not in STRUCTURAL_STATUSES
    )

    maintenance_vids = set()
    for m in maintenances:
        if (_get(m, "status") or "").upper() in TERMINAL_MAINTENANCE_STATUSES:
            continue
        vid = _get(m, "vehicle_id")
        if vid in structural:
            continue
        m_start = parse_datetime_utc(_get(m, "start_datetime"))
        if m_start and m_start <= now < _maintenance_end(m):
            maintenance_vids.add(vid)

    rented_vids = set()
    reserved_vids = set()
    for r in reservations:
        st = (_get(r, "status") or "").upper()
        if st not in BLOCKING_RESERVATION_STATUSES:
            continue
        vid = _get(r, "vehicle_id")
        if vid in structural or vid in maintenance_vids:
            continue
        r_start = parse_datetime_utc(_get(r, "start_datetime"))
        r_end = parse_datetime_utc(_get(r, "end_datetime"))
        if not r_start or not r_end:
            continue
        if r_start <= now < r_end:
            # Window contains now -> car is out (no separate pickup step).
            rented_vids.add(vid)
        elif now < r_start:
            # Upcoming booking -> surfaced as RESERVED.
            reserved_vids.add(vid)

    reserved_vids -= rented_vids
    return rented_vids, reserved_vids, maintenance_vids, total_vehicles


def compute_fleet_counts_rows(vehicles, reservations, maintenances, now=None):
    rented, reserved, maint, total = compute_fleet_sets_rows(
        vehicles, reservations, maintenances, now)
    return {
        "total_vehicles": total,
        "available": max(0, total - len(rented) - len(reserved) - len(maint)),
        "reserved": len(reserved),
        "rented": len(rented),
        "maintenance": len(maint),
    }


def effective_statuses_rows(vehicles, reservations, maintenances, now=None):
    """{vehicle_id: effective status} for every vehicle."""
    rented, reserved, maint, _total = compute_fleet_sets_rows(
        vehicles, reservations, maintenances, now)
    out = {}
    for v in vehicles:
        vid = _get(v, "id")
        out[str(vid)] = effective_status(_get(v, "status"), vid, rented, reserved, maint)
    return out


def next_local_midnight(now=None, tz_name="Africa/Casablanca"):
    """The next local-midnight in ``tz_name`` strictly after ``now`` (aware UTC).
    Used for dashboard period (today/week/month) rollover."""
    from zoneinfo import ZoneInfo
    tz = ZoneInfo(tz_name)
    now = (now or datetime.now(timezone.utc)).astimezone(tz)
    local_midnight = (now + timedelta(days=1)).replace(
        hour=0, minute=0, second=0, microsecond=0)
    return local_midnight.astimezone(timezone.utc)


def next_boundary_rows(reservations, maintenances, now=None,
                       include_midnight=False, tz_name="Africa/Casablanca"):
    """The earliest FUTURE instant (strictly ``> now``) at which some vehicle's
    effective status — or (when ``include_midnight``) a dashboard period card —
    can change with no user action: the reservation/maintenance interval edges,
    plus the next local midnight. Returns an aware-UTC datetime, or None when
    nothing is pending. Half-open semantics: an edge at exactly ``now`` has
    already taken effect and is not returned.
    """
    now = now or datetime.now(timezone.utc)
    cands = []

    for r in reservations:
        if (_get(r, "status") or "").upper() not in BLOCKING_RESERVATION_STATUSES:
            continue
        for key in ("start_datetime", "end_datetime"):
            dt = parse_datetime_utc(_get(r, key))
            if dt is not None and dt > now:
                cands.append(dt)

    for m in maintenances:
        if (_get(m, "status") or "").upper() in TERMINAL_MAINTENANCE_STATUSES:
            continue
        start = parse_datetime_utc(_get(m, "start_datetime"))
        if start is not None and start > now:
            cands.append(start)
        end = _maintenance_end(m)
        if end != FAR_FUTURE and end > now:
            cands.append(end)

    if include_midnight:
        cands.append(next_local_midnight(now, tz_name))

    return min(cands) if cands else None


# ── session adapters (unchanged public API) ────────────────────────────────
def _rows_from_session(session):
    from app.models.vehicle import LocalVehicle
    from app.models.reservation import LocalReservation
    from app.models.maintenance import LocalMaintenance
    return (
        session.query(LocalVehicle).all(),
        session.query(LocalReservation).all(),
        session.query(LocalMaintenance).all(),
    )


def compute_fleet_sets(session, now=None):
    """Return (rented_vids, reserved_vids, maintenance_vids, total_vehicles)
    from the local SQLite session (thin adapter over the pure core)."""
    vehicles, reservations, maintenances = _rows_from_session(session)
    return compute_fleet_sets_rows(vehicles, reservations, maintenances, now)


def compute_fleet_counts(session, now=None):
    """Canonical fleet breakdown — mutually exclusive, sums to total_vehicles."""
    vehicles, reservations, maintenances = _rows_from_session(session)
    return compute_fleet_counts_rows(vehicles, reservations, maintenances, now)


def effective_status(v_status, vehicle_id, rented_vids, reserved_vids, maintenance_vids):
    """Resolve one vehicle's effective status given the disjoint sets."""
    if (v_status or "").upper() in STRUCTURAL_STATUSES:
        return v_status
    if vehicle_id in maintenance_vids:
        return EFFECTIVE_MAINTENANCE
    if vehicle_id in rented_vids:
        return EFFECTIVE_RENTED
    if vehicle_id in reserved_vids:
        return EFFECTIVE_RESERVED
    return EFFECTIVE_AVAILABLE

File: mobile/app/src/main/java/com/example/data/fleet/FleetStatus.kt
package com.example.data.fleet

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Kotlin adapter of the NORMATIVE fleet-status specification
 * (`shared/fleet_status_reference.py` + `shared/fleet_status_cases.json`) —
 * the mobile mirror of `desktop/app/utils/fleet_status.py` and
 * `backend/app/services/fleet_status.py`.
 *
 * This file re-implements NO business rule of its own: it is a mechanical port
 * of the shared reference, and `FleetStatusParityTest` drives it with the
 * shared cross-runtime vectors so it can never drift from Desktop/Backend.
 *
 * PRECEDENCE (mutually exclusive):
 *   SOLD / INACTIVE  (structural) > MAINTENANCE > RENTED > RESERVED > AVAILABLE
 *
 * RENTED is time-derived: a blocking reservation (RESERVED or ACTIVE) whose
 * window contains `now` means the car is physically out (no separate pickup
 * step). RESERVED means a blocking reservation is still UPCOMING (`now < start`).
 *
 * INTERVAL RULE: half-open [start, end). Occupied for `start <= now < end`;
 * exactly at `end` the vehicle is free again.
 *
 * All time is epoch-milliseconds UTC (the codebase's existing convention —
 * `System.currentTimeMillis()` / `SimpleDateFormat` with an explicit TZ — so
 * no `java.time` / desugaring dependency is added).
 *
 * NAIVE-DATETIME POLICY (the ONE policy, product-wide): an ISO string carrying
 * no offset is BUSINESS-LOCAL wall time (Africa/Casablanca), never UTC. This is
 * the contract written in `shared/money_time.to_business` and mirrored by
 * `shared/fleet_status_reference._parse`, `shared/revenue_reference`,
 * `desktop/app/utils/datetime_utils.parse_datetime_utc` and
 * `backend/app/services/fleet_status`. Reading a naive value as UTC shifted
 * every such row by the Casablanca offset and could put Mobile in a different
 * bucket than Desktop/Backend for the SAME row at the SAME instant; the
 * `naive_*` vectors in `shared/fleet_status_cases.json` pin this down.
 */
object FleetStatus {

    const val AVAILABLE = "AVAILABLE"
    const val RESERVED = "RESERVED"
    const val RENTED = "RENTED"
    const val MAINTENANCE = "MAINTENANCE"

    val CASABLANCA: TimeZone = TimeZone.getTimeZone("Africa/Casablanca")

    private val STRUCTURAL = setOf("SOLD", "INACTIVE")
    private val TERMINAL_MAINTENANCE = setOf("COMPLETED", "CANCELLED")
    private val BLOCKING_RESERVATION = setOf("ACTIVE", "RESERVED")
    private const val FAR_FUTURE = Long.MAX_VALUE

    data class VehicleRow(val id: String, val status: String?)
    data class ReservationRow(
        val vehicleId: String,
        val status: String?,
        val startIso: String?,
        val endIso: String?,
        val totalAmount: Double = 0.0,
        val numDays: Int = 1,
        val cancellationReason: String? = null,
        val cancelledAtIso: String? = null,
    )

    data class MaintenanceRow(
        val vehicleId: String,
        val status: String?,
        val startIso: String?,
        val expectedEndIso: String?,
        val actualEndIso: String?,
    )

    // ── canonical parser (mirror of parse_datetime_utc) ────────────────────
    // A format WITHOUT "XXX" has no offset in the text, so the parser's own
    // timeZone decides the instant: that is the naive case, and it resolves to
    // BUSINESS_TZ (see the class docstring), not UTC.
    private val ISO_FORMATS = listOf(
        "yyyy-MM-dd'T'HH:mm:ssXXX",   // explicit offset -> as written
        "yyyy-MM-dd'T'HH:mm:ss",      // naive -> business-local
        "yyyy-MM-dd HH:mm:ss",        // SQLite round-trip -> business-local
        "yyyy-MM-dd'T'HH:mm",         // naive -> business-local
        "yyyy-MM-dd",                 // naive date -> business-local midnight
    )

    fun parseUtcMillis(value: String?): Long? {
        if (value.isNullOrBlank()) return null
        var s = value.trim().replace("z", "Z")
        // normalise "...Z" to "...+00:00" so XXX can parse it
        if (s.endsWith("Z")) s = s.dropLast(1) + "+00:00"
        // drop fractional seconds: "12:00:00.123456+00:00" -> "12:00:00+00:00"
        val dot = s.indexOf('.')
        if (dot >= 0) {
            var j = dot + 1
            while (j < s.length && s[j].isDigit()) j++
            s = s.substring(0, dot) + s.substring(j)
        }
        for (fmt in ISO_FORMATS) {
            try {
                val sdf = SimpleDateFormat(fmt, Locale.US)
                sdf.isLenient = false
                // No offset in the pattern => naive => business-local wall time.
                if (!fmt.contains("XXX")) sdf.timeZone = CASABLANCA
                return sdf.parse(s)?.time
            } catch (_: Exception) {
            }
        }
        return null
    }

    private fun norm(s: String?): String = (s ?: "").trim().uppercase(Locale.US)

    private fun maintenanceEndMillis(m: MaintenanceRow): Long =
        parseUtcMillis(m.actualEndIso) ?: parseUtcMillis(m.expectedEndIso) ?: FAR_FUTURE

    // ── effective status ─────────────────────────────────────────────────
    fun effectiveStatuses(
        vehicles: List<VehicleRow>,
        reservations: List<ReservationRow>,
        maintenances: List<MaintenanceRow>,
        nowMillis: Long,
    ): Map<String, String> {
        val result = LinkedHashMap<String, String>()
        val structural = HashSet<String>()
        for (v in vehicles) {
            val st = norm(v.status)
            if (st in STRUCTURAL) {
                result[v.id] = st
                structural.add(v.id)
            } else {
                result[v.id] = AVAILABLE
            }
        }

        val maintVids = HashSet<String>()
        for (m in maintenances) {
            if (m.vehicleId in structural || m.vehicleId !in result) continue
            if (norm(m.status) in TERMINAL_MAINTENANCE) continue
            val start = parseUtcMillis(m.startIso) ?: continue
            if (start <= nowMillis && nowMillis < maintenanceEndMillis(m)) {
                maintVids.add(m.vehicleId)
            }
        }

        val rentedVids = HashSet<String>()
        val reservedVids = HashSet<String>()
        for (r in reservations) {
            if (r.vehicleId in structural || r.vehicleId in maintVids || r.vehicleId !in result) continue
            val st = norm(r.status)
            if (st !in BLOCKING_RESERVATION) continue
            val start = parseUtcMillis(r.startIso) ?: continue
            val end = parseUtcMillis(r.endIso) ?: continue
            if (start <= nowMillis && nowMillis < end) {
                // Window contains now -> car is out (no separate pickup step).
                rentedVids.add(r.vehicleId)
            } else if (nowMillis < start) {
                // Upcoming booking -> surfaced as RESERVED.
                reservedVids.add(r.vehicleId)
            }
        }
        reservedVids.removeAll(rentedVids)

        for (id in result.keys.toList()) {
            if (id in structural) continue
            result[id] = when (id) {
                in maintVids -> MAINTENANCE
                in rentedVids -> RENTED
                in reservedVids -> RESERVED
                else -> AVAILABLE
            }
        }
        return result
    }

    fun fleetCounts(
        vehicles: List<VehicleRow>,
        reservations: List<ReservationRow>,
        maintenances: List<MaintenanceRow>,
        nowMillis: Long,
    ): Map<String, Int> {
        val statuses = effectiveStatuses(vehicles, reservations, maintenances, nowMillis)
        var available = 0; var reserved = 0; var rented = 0; var maintenance = 0
        for (s in statuses.values) when (s) {
            AVAILABLE -> available++
            RESERVED -> reserved++
            RENTED -> rented++
            MAINTENANCE -> maintenance++
        }
        val total = available + reserved + rented + maintenance
        return mapOf(
            "total_vehicles" to total,
            "available" to available,
            "reserved" to reserved,
            "rented" to rented,
            "maintenance" to maintenance,
        )
    }

    // ── next temporal boundary ───────────────────────────────────────────
    /**
     * Earliest FUTURE instant (strictly `> nowMillis`) at which some vehicle's
     * effective status could change purely because time advanced — the
     * reservation / maintenance interval edges, and (when [includeMidnight])
     * the next local-midnight in [zone] for dashboard period rollover.
     * Returns null when nothing is pending.
     */
    fun nextBoundaryMillis(
        reservations: List<ReservationRow>,
        maintenances: List<MaintenanceRow>,
        nowMillis: Long,
        includeMidnight: Boolean = false,
        zone: TimeZone = CASABLANCA,
    ): Long? {
        var best: Long? = null
        fun consider(t: Long?) {
            if (t != null && t > nowMillis && (best == null || t < best!!)) best = t
        }

        for (r in reservations) {
            if (norm(r.status) !in BLOCKING_RESERVATION) continue
            consider(parseUtcMillis(r.startIso))
            consider(parseUtcMillis(r.endIso))
        }
        for (m in maintenances) {
            if (norm(m.status) in TERMINAL_MAINTENANCE) continue
            consider(parseUtcMillis(m.startIso))
            val end = maintenanceEndMillis(m)
            if (end != FAR_FUTURE) consider(end)
        }
        if (includeMidnight) consider(nextMidnightMillis(nowMillis, zone))
        return best
    }

    // ── dashboard period overview (mirror of dashboard_cache.compute_overview_rows) ──
    data class PeriodOverview(
        val totalVehicles: Int, val available: Int, val rented: Int,
        val reserved: Int, val maintenance: Int,
        val todayBookings: Int, val weekBookings: Int, val monthBookings: Int, val yearBookings: Int,
        val todayRevenue: Double, val weekRevenue: Double, val monthRevenue: Double, val yearRevenue: Double,
    )

    data class PeriodRanges(
        val today: LongRange, val week: LongRange, val month: LongRange, val year: LongRange,
    )

    /** Africa/Casablanca-local [today, tomorrow), [weekStart Mon, +7d),
     *  [monthStart, nextMonth), [yearStart, nextYear) as epoch-millis bounds. */
    private fun periodBounds(nowMillis: Long, zone: TimeZone): PeriodRanges {
        fun cal() = Calendar.getInstance(zone, Locale.US).apply {
            firstDayOfWeek = Calendar.MONDAY; time = Date(nowMillis)
            set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
        }
        val todayStart = cal().timeInMillis
        val todayEnd = cal().apply { add(Calendar.DAY_OF_YEAR, 1) }.timeInMillis
        val weekStart = cal().apply {
            val dow = (get(Calendar.DAY_OF_WEEK) + 5) % 7   // Mon=0..Sun=6
            add(Calendar.DAY_OF_YEAR, -dow)
        }.timeInMillis
        val weekEnd = weekStart + 7L * 86_400_000L
        val monthStart = cal().apply { set(Calendar.DAY_OF_MONTH, 1) }.timeInMillis
        val monthEnd = cal().apply { set(Calendar.DAY_OF_MONTH, 1); add(Calendar.MONTH, 1) }.timeInMillis
        val yearStart = cal().apply { set(Calendar.MONTH, Calendar.JANUARY); set(Calendar.DAY_OF_MONTH, 1) }.timeInMillis
        val yearEnd = cal().apply {
            set(Calendar.MONTH, Calendar.JANUARY); set(Calendar.DAY_OF_MONTH, 1); add(Calendar.YEAR, 1)
        }.timeInMillis
        return PeriodRanges(
            todayStart until todayEnd, weekStart until weekEnd,
            monthStart until monthEnd, yearStart until yearEnd,
        )
    }

    fun dashboardOverview(
        vehicles: List<VehicleRow>,
        reservations: List<ReservationRow>,
        maintenances: List<MaintenanceRow>,
        nowMillis: Long,
        zone: TimeZone = CASABLANCA,
    ): PeriodOverview {
        val counts = fleetCounts(vehicles, reservations, maintenances, nowMillis)
        val p = periodBounds(nowMillis, zone)

        // booking COUNT stays anchored to the start date (a rental "made in
        // September" by when it started).
        var tB = 0; var wB = 0; var mB = 0; var yB = 0
        for (r in reservations) {
            // Mirror shared.revenue_reference.rentals_started_between: a
            // maintenance-interrupted rental still "started" and still counts;
            // any other CANCELLED does not.
            if (norm(r.status) == "CANCELLED" && norm(r.cancellationReason) != "MAINTENANCE") continue
            val start = parseUtcMillis(r.startIso) ?: continue
            if (start in p.today) tB++; if (start in p.week) wB++
            if (start in p.month) mB++; if (start in p.year) yB++
        }

        // revenue = the ONE pro-rata engine (shared/revenue_reference.py).
        val rentals = reservations.map {
            RevenueEngine.Rental(
                status = it.status,
                startMillis = parseUtcMillis(it.startIso),
                numDays = it.numDays,
                totalPrice = java.math.BigDecimal(it.totalAmount.toString()),
                cancellationReason = it.cancellationReason,
                cancelledAtMillis = parseUtcMillis(it.cancelledAtIso),
                endMillis = parseUtcMillis(it.endIso),
            )
        }
        fun rev(name: String): Double {
            val (f, t) = RevenueEngine.namedPeriodBounds(name, nowMillis, zone)
            return RevenueEngine.revenueBetween(rentals, f, t, nowMillis, zone)
        }
        return PeriodOverview(
            counts["total_vehicles"]!!, counts["available"]!!, counts["rented"]!!,
            counts["reserved"]!!, counts["maintenance"]!!,
            tB, wB, mB, yB,
            rev("today"), rev("week"), rev("month"), rev("year"),
        )
    }

    /** The next local midnight in [zone] strictly after [nowMillis]. */
    fun nextMidnightMillis(nowMillis: Long, zone: TimeZone = CASABLANCA): Long {
        val cal = Calendar.getInstance(zone, Locale.US)
        cal.time = Date(nowMillis)
        cal.add(Calendar.DAY_OF_YEAR, 1)
        cal.set(Calendar.HOUR_OF_DAY, 0)
        cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }
}
